RecoCTPPS/PixelLocal/src/RPixPlaneCombinatoryTracking.cc

```diff
@@ -0,0 +1,483 @@
+            std::pair<double,double> residuals = std::make_pair(xResidual,yResidual);
+            TMatrixD globalError = hit.globalError;
+            std::pair<double,double> pulls = std::make_pair(xResidual/TMath::Sqrt(globalError[0][0]),yResidual/TMath::Sqrt(globalError[1][1]));
+            delete fittedRecHit;
+            fittedRecHit = new CTPPSPixelLocalTrack::CTPPSPixelFittedRecHit(hit.recHit, pointOnDet, residuals, pulls);
+            fittedRecHit->setIsRealHit(true);
+          }
+        }
+      }
+      else{
+        fittedRecHit->setIsRealHit(false);
+        fittedRecHit->setGlobalCoordinates(pointOnDet);
+      }
+
+      bestTrack.addHit(tmpPlaneId, *fittedRecHit);
+  
+    }
+
+    localTrackVector_.push_back(bestTrack);
+
+    int pointForTracking = 0;
+    int pointOnTrack = 0;
+
+    if(verbosity_>=1){
+      for(const auto & planeHits : bestTrack.getHits()){
+        for(const auto & fittedhit : planeHits){
+          if(fittedhit.getIsUsedForFit()) ++pointForTracking;
+          if(fittedhit.getIsRealHit()) ++pointOnTrack;
+        }
+      }
+      std::cout<<"Best track has "<<pointForTracking<<" points used for the fit and " << pointOnTrack<<" points belonging to the track\n";
+    }
+
+  } //close of the while loop on all the hits
+
+  return;
+
+}
+
+//------------------------------------------------------------------------------------------------//
+
+CTPPSPixelLocalTrack RPixPlaneCombinatoryTracking::fitTrack(std::vector<RPixDetPatternFinder::PointInPlane> pointList){
+  
+  uint32_t numberOfPoints = pointList.size();
+  TVectorD xyCoordinates(2*numberOfPoints);
+  TMatrixD varianceMatrix(2*numberOfPoints,2*numberOfPoints);
+  TMatrixD zMatrix(2*numberOfPoints,4);
+  
+  //The matrices and vector xyCoordinates, varianceMatrix and varianceMatrix are built from the points
+  uint32_t hitCounter=0;
+  for(const auto & hit : pointList){
+    CLHEP::Hep3Vector globalPoint  = hit.globalPoint    ;
+    xyCoordinates[hitCounter]      = globalPoint.x()    ;
+    xyCoordinates[hitCounter+1]    = globalPoint.y()    ;
+    zMatrix      [hitCounter][0]   = 1.                 ;
+    zMatrix      [hitCounter][2]   = globalPoint.z()-z0_;
+    zMatrix      [hitCounter+1][1] = 1.                 ;
+    zMatrix      [hitCounter+1][3] = globalPoint.z()-z0_;
+
+    TMatrixD globalError = hit.globalError;
+    varianceMatrix[hitCounter][hitCounter]     = globalError[0][0];
+    varianceMatrix[hitCounter][hitCounter+1]   = globalError[0][1];
+    varianceMatrix[hitCounter+1][hitCounter]   = globalError[1][0];
+    varianceMatrix[hitCounter+1][hitCounter+1] = globalError[1][1];
+    
+    hitCounter+=2;
+  }
+
+  //for having the real point variance matrix, varianceMatrix need to be inverted
+  try{
+    varianceMatrix.Invert();
+  }
+  catch (cms::Exception &e){
+    edm::LogError("RPixPlaneCombinatoryTracking") << "Error in RPixPlaneCombinatoryTracking::fitTrack -> "
+    << "Point variance matrix is singular, skipping.";
```

Comment:
is this rare enough (less than once per 100 events)?
If not, logError is inappropriate.

Reply:
Yes, it is rare.